packages/@lwc/engine-core/src/framework/base-lightning-element.ts

```diff
@@ -300,6 +300,60 @@ function warnIfInvokedDuringConstruction(vm: VM, methodOrPropName: string) {
     }
 }
 
+// List of properties on ElementInternals that are formAssociated can be found in the spec:
+// https://html.spec.whatwg.org/multipage/custom-elements.html#form-associated-custom-elements
+const formAssociatedProps = new Set([
+    'setFormValue',
+    'form',
+    'setValidity',
+    'willValidate',
+    'validity',
+    'validationMessage',
+    'checkValidity',
+    'reportValidity',
+    'labels',
+]);
+
+// Wrap all ElementInternal objects in a proxy to prevent form association when `formAssociated` is not set on an LWC.
+// This is needed because the 1UpgradeableConstructor1 always sets `formAssociated=true`, which means all
+// ElementInternal objects will have form-associated properties set when an LWC is placed in a form.
+// We are doing this to guard against customers taking a dependency on form elements being associated to ElementInternals
+// when 'formAssociated' has not been set on the LWC.
+function createElementInternalsProxy(
+    elementInternals: ElementInternals,
+    isFormAssociated: boolean
+) {
+    const elementInternalsProxy = new Proxy(elementInternals, {
+        set(target, propertyKey, newValue) {
+            const key = typeof propertyKey === 'symbol' ? propertyKey.description : propertyKey;
+            if (key && formAssociatedProps.has(key) && !isFormAssociated) {
+                //Note this error message mirrors Chrome and Firefox error messages, in Safari the error is slightly different.
+                throw new DOMException(
+                    `Failed to execute '${key}' on 'ElementInternals': The target element is not a form-associated custom element.`
+                );
+            }
+            return Reflect.set(target, propertyKey, newValue);
+        },
+        get(target, propertyKey) {
+            const key = typeof propertyKey === 'symbol' ? propertyKey.description : propertyKey;
+            if (key && formAssociatedProps.has(key) && !isFormAssociated) {
+                //Note this error message mirrors Chrome and Firefox error messages, in Safari the error is slightly different.
+                throw new DOMException(
+                    `Failed to execute '${key}' on 'ElementInternals': The target element is not a form-associated custom element.`
+                );
+            }
+            const internalsPropertyValue = Reflect.get(target, propertyKey);
+            // Bind the property value to the target so that function invocations are called with the
+            // correct context ('this' value).
+            return typeof internalsPropertyValue === 'function'
+                ? internalsPropertyValue.bind(target)
+                : internalsPropertyValue;
```

Comment:
This is so function invocations like `ElementInternals.setFormValue('on', 'checked')` is able to be invoked with the correct context (`ElementInternals` instance).

Reply:
TIL. I assumed Proxies just handled the binding for you when you do `foo.bar()`.

Reply:
I looked into it, and it seems [proxies should handle the `this` properly](https://codepen.io/nolanlawson-the-selector/pen/mdaKXWJ).

I think what's happening here is that the browser has some internal logic that is throwing `TypeError: Illegal invocation` when using the proxy. The `bind` is a nice workaround for this, although technically it does change the semantics. (E.g. `const foo = internals.checkValidity && foo()` should not work.) However, this is definitely an edge case.

[Googling didn't turn up much](https://stackoverflow.com/questions/42496414/illegal-invocation-error-using-es6-proxy-and-node-js), but yeah, seems like `bind` is the typical workaround.